File: packages/server/src/decay.ts
/**
 * Decay computation utilities
 */

import type { DecayModel, Pheromone } from "./types.js";

/**
 * Compute the current intensity of a pheromone after decay
 */
export function computeIntensity(pheromone: Pheromone, now: number): number {
  const elapsed = now - pheromone.last_reinforced_at;

  if (elapsed <= 0) {
    return pheromone.initial_intensity;
  }

  switch (pheromone.decay_model.type) {
    case "exponential": {
      const halfLife = pheromone.decay_model.half_life_ms;
      return pheromone.initial_intensity * Math.pow(0.5, elapsed / halfLife);
    }

    case "linear": {
      const rate = pheromone.decay_model.rate_per_ms;
      return Math.max(0, pheromone.initial_intensity - rate * elapsed);
    }

    case "step": {
      const steps = pheromone.decay_model.steps;
      // Find the applicable step (steps should be sorted by at_ms)
      for (let i = steps.length - 1; i >= 0; i--) {
        if (elapsed >= steps[i].at_ms) {
          // Invariant (SPECIFICATION.md §4.3): computed intensity never
          // exceeds initial_intensity, so a step is clamped into
          // [0, initial_intensity] like the other decay models.
          return Math.max(0, Math.min(steps[i].intensity, pheromone.initial_intensity));
        }
      }
      return pheromone.initial_intensity;
    }

    case "immortal":
      return pheromone.initial_intensity;

    default:
      return 0;
  }
}

/**
 * Check if a pheromone has evaporated below its floor
 */
export function isEvaporated(pheromone: Pheromone, now: number): boolean {
  return computeIntensity(pheromone, now) < pheromone.ttl_floor;
}

/**
 * Get the default decay model
 */
export function defaultDecay(): DecayModel {
  return { type: "exponential", half_life_ms: 300000 }; // 5 minutes
}

/**
 * Estimate time until pheromone evaporates
 */
export function timeToEvaporation(pheromone: Pheromone, now: number): number | null {
  const currentIntensity = computeIntensity(pheromone, now);

  if (currentIntensity <= pheromone.ttl_floor) {
    return 0; // Already evaporated
  }

  switch (pheromone.decay_model.type) {
    case "exponential": {
      // Solve: ttl_floor = current * 0.5^(t/halfLife)
      // t = halfLife * log2(current / ttl_floor)
      const halfLife = pheromone.decay_model.half_life_ms;
      const ratio = currentIntensity / pheromone.ttl_floor;
      return halfLife * Math.log2(ratio);
    }

    case "linear": {
      // Solve: ttl_floor = current - rate * t
      // t = (current - ttl_floor) / rate
      const rate = pheromone.decay_model.rate_per_ms;
      if (rate <= 0) return null;
      return (currentIntensity - pheromone.ttl_floor) / rate;
    }

    case "step": {
      // Find next step below threshold
      const steps = pheromone.decay_model.steps;
      const elapsed = now - pheromone.last_reinforced_at;
      for (const step of steps) {
        if (step.at_ms > elapsed && step.intensity < pheromone.ttl_floor) {
          return step.at_ms - elapsed;
        }
      }
      return null; // May never evaporate with given steps
    }

    case "immortal":
      return null; // Never evaporates
  }
}

File: packages/client-python/src/sbp/decay.py
"""
Decay computation utilities
"""
import math
from typing import Optional
from sbp.types import DecayModel, Pheromone

def compute_intensity(pheromone: Pheromone, now: int) -> float:
    """Compute the current intensity of a pheromone after decay"""
    elapsed = now - pheromone.last_reinforced_at

    if elapsed <= 0:
        return pheromone.initial_intensity

    decay = pheromone.decay_model

    if decay.type == "exponential":
        return pheromone.initial_intensity * math.pow(0.5, elapsed / decay.half_life_ms) # type: ignore

    elif decay.type == "linear":
        rate = decay.rate_per_ms # type: ignore
        return max(0.0, pheromone.initial_intensity - rate * elapsed)

    elif decay.type == "step":
        steps = decay.steps # type: ignore
        # Find the applicable step (steps should be sorted by at_ms)
        for step in reversed(steps):
            if elapsed >= step["at_ms"]:
                # Invariant (SPECIFICATION.md §4.3): computed intensity never
                # exceeds initial_intensity, so a step is clamped into
                # [0, initial_intensity] like the other decay models.
                return max(0.0, min(step["intensity"], pheromone.initial_intensity))
        return pheromone.initial_intensity

    elif decay.type == "immortal":
        return pheromone.initial_intensity

    return 0.0

def is_evaporated(pheromone: Pheromone, now: int) -> bool:
    """Check if a pheromone has evaporated below its floor"""
    return compute_intensity(pheromone, now) < pheromone.ttl_floor
